### PR TITLE
Add configurable recursive field handling: mode option, depth limiting, and comprehensive testing

### PR DESCRIPTION
**Pull request by bumfo**
## Summary

Adds **configurable recursive field handling** to the protobuf connector with new `recursive.fields.mode` option, depth limiting, and comprehensive testing. Users can now choose how recursive protobuf schemas are handled: fail, drop, mock as binary, or allow with `RecursiveStructType`.

**This is a major new feature** - before this PR, recursive handling was hardcoded. Now users have full control via configuration.

## What Existed Before (origin/master)

**ProtobufOptions**:
- `parseMode` - PERMISSIVE vs FAILFAST
- `recursiveFieldMaxDepth` - existed but **not used anywhere**
- NO `recursive.fields.mode` option

**RecursiveSchemaConverters**:
- `toSqlTypeWithRecursionMocking()` - hardcoded to mock as BinaryType
- `toSqlTypeWithTrueRecursion()` - hardcoded to use RecursiveStructType
- NO unified API, NO mode-based configuration, NO depth limiting

**Tests**: 0 tests for recursive field modes

**Result**: Users had no control over recursive field handling behavior.

### 1. NEW: recursive.fields.mode Configuration Option

**User-facing configuration** to control recursive field handling:

```scala
Map("recursive.fields.mode" -> "drop")      // Drop recursive fields entirely
Map("recursive.fields.mode" -> "binary")    // Mock as BinaryType
Map("recursive.fields.mode" -> "fail")      // Fail on recursive schemas
Map("recursive.fields.mode" -> "recursive") // Allow RecursiveStructType
```

### 2. NEW: Working Depth Limiting

**Connected depth config to actual implementation**:

```scala
Map("recursive.fields.max.depth" -> "0")  // No recursive fields allowed
Map("recursive.fields.max.depth" -> "3")  // Allow 3 recursions, then drop
Map("recursive.fields.max.depth" -> "-1") // Default (parser-dependent)
```

**Depth semantics**:
- `-1`: Default (recursive for WireFormat parser, fail for others)
- `0`: No recursions (drop/fail/mock on first recursion)
- `1-10`: Allow N recursions (drop when depth > N)

**Example** with depth=3 and A↔B mutual recursion:
```
A(0) → B(0) → A(1) → B(2) → A(3) → B(4>3, DROPPED)
Result: A → B → A → B → A (primitives only)
```

### 3. NEW: RecursionMode Enum (Type Safety)

```scala
sealed trait RecursionMode
object RecursionMode {
  case object Fail extends RecursionMode
  case object Drop extends RecursionMode
  case object MockAsBinary extends RecursionMode
  case object AllowRecursive extends RecursionMode
}
```

### 4. NEW: Unified toSqlType() API

**Single entry point** respecting mode + depth configuration:

```scala
def toSqlType(
    descriptor: Descriptor,
    recursiveFieldsMode: RecursionMode,
    recursiveFieldMaxDepth: Int,
    allowRecursion: Boolean,
    enumAsInt: Boolean = false): DataType
```

### 5. NEW: Configuration Precedence Logic

**depth=-1 (default)**: mode="" uses parser default, explicit modes override
**depth=0 (no recursions)**: mode="" drops, explicit modes override
**depth>=1 (N recursions)**: mode="" drops when depth > N

**Important**: `depth=-1` + `mode="drop"/"binary"/"fail"` = no recursions (same as `depth=0`)

### 6. NEW: Comprehensive Test Suite (516 lines)

**RecursiveFieldsModeSpec.scala** with 19 scenarios:
- Self-recursion tests (Recursive message)
- Mutual recursion tests (MutualA ↔ MutualB)
- Depth limit tests (depth=1,2,3)
- Mode precedence tests
- Edge cases (depth=-1 + explicit modes)

**Test results**: 101/101 passing

### 7. FIXED: RecursiveStructType Serialization

- Use `InternalRow.fromSeq()` for proper circular reference handling
- Match Spark `StructType` Kryo registration

### 8. NEW: Complete Documentation

**ProtobufOptions.scala** (+207 lines): Complete configuration guide with precedence matrix
**CLAUDE.md** (+40 lines): Usage examples
**Python** (+36 lines): Complete docstring

## Files Changed (9 files, +1,123/-104 lines)

- `RecursiveSchemaConverters.scala` +312 - RecursionMode enum, unified API, depth limiting
- `RecursiveFieldsModeSpec.scala` +516 (new) - Test suite (19 scenarios)
- `ProtobufOptions.scala` +207 - Mode option, enum parsing, docs
- `RecursiveStructType.scala` +33 - Serialization fix
- `RecursiveStructTypeDifferencesSpec.scala` +65 (new) - Compatibility tests
- `CLAUDE.md` +40 - Usage examples
- `python/spark_protobuf/functions.py` +36 - Docstring
- `ProtobufDataToCatalyst.scala` +16 - Use mode config
- 9 test files - Use RecursionMode enum

### Before This PR (Hardcoded)

```scala
// Only 2 options, both hardcoded:
val schema1 = RecursiveSchemaConverters.toSqlTypeWithRecursionMocking(descriptor)
val schema2 = RecursiveSchemaConverters.toSqlTypeWithTrueRecursion(descriptor)
// NO WAY to drop fields, fail on recursion, or limit depth
```

### After This PR (User-Configurable)

```scala
// Drop recursive fields
df.select(from_protobuf($"data", "Message", descBytes,
  Map("recursive.fields.mode" -> "drop")))

// Allow 3 recursions, then drop
df.select(from_protobuf($"data", "Message", descBytes,
  Map("recursive.fields.max.depth" -> "3")))

// Unlimited recursion (RecursiveStructType)
df.select(from_protobuf($"data", "Message", descBytes,
  Map("recursive.fields.mode" -> "recursive")))
```

## Testing

```bash
$ sbt unitTests
[info] Total number of tests run: 101
[info] Tests: succeeded 101, failed 0
[info] All tests passed.
```

## Migration Notes

**No breaking changes** - this is a pure addition. Existing code works exactly as before.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

Co-Authored-By: Claude <noreply@anthropic.com>

### CODE REVIEW

File: core/src/main/scala/fastproto/RecursiveSchemaConverters.scala [hunk 1]
```diff
@@ -85,71 +389,122 @@ object RecursiveSchemaConverters {
...
+                                   cycleDepth, maxRecursiveDepth))
+          }
         } else {
-          // Not recursive yet, continue normal conversion
-          convertMessageType(messageDescriptor, visitedTypes.clone(), enumAsInt)
+          // Not recursive - check if we're inside a cycle
+          val nextDepth = if (recursiveDepth > 0) {
+            // We're inside a cycle (visiting C in A→B→A→C pattern)
+            // Increment depth for this new field
+            recursiveDepth + 1
```
[the remark this answers is not in the document]
Comment (bumfo) [reply core/src/main/scala/fastproto/RecursiveSchemaConverters.scala]: document updated

**Comment by bumfo** [comment]
@codex review

**Comment by bumfo** [comment]
@codex review

**Comment by bumfo** [comment]
@codex review
